src/Native/Unix/System.Security.Cryptography.Native/pal_ssl.cpp

```diff
@@ -44,26 +63,62 @@ extern "C" SSL_CTX* CryptoNative_SslCtxCreate(SSL_METHOD* method)
     return ctx;
 }
 
+static uint32_t ParseOpenSSLVersion(const char* string_version)
+{
+    const char* start = string_version + sizeof("OpenSSL");
+    char* end;
+    uint32_t major = static_cast<uint32_t>(strtoul(start, &end, 10));
+    start = end + 1;
+    uint32_t minor = static_cast<uint32_t>(strtoul(start, &end, 10));
+    start = end + 1;
+    uint32_t update = static_cast<uint32_t>(strtoul(start, &end, 10));
+
+    uint32_t version = (major << 28) + (minor << 20) + (update << 12);
+    return version;
+}
+
 /*
 Openssl supports setting ecdh curves by default from version 1.1.0.
 For lower versions, this is the recommended approach.
 Returns 1 on success, 0 on failure.
 */
 static long TrySetECDHNamedCurve(SSL_CTX* ctx)
 {
-	long result = 0;
-#ifdef SSL_CTX_set_ecdh_auto
-	result = SSL_CTX_set_ecdh_auto(ctx, 1);
-#else
-	EC_KEY *ecdh = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
-	if (ecdh != nullptr)
-	{
-		result = SSL_CTX_set_tmp_ecdh(ctx, ecdh);
-		EC_KEY_free(ecdh);
-	}
+#ifdef NEED_OPENSSL_1_0
+    const char* string_version = CryptoNative_SSLEayVersion();
+    uint32_t version = ParseOpenSSLVersion(string_version);
+    long result = 0;
+
+    if (version >= OPENSSL_VERSION_1_1_0_RTM)
+    {
+        // OpenSSL 1.1+ automatically set up ECDH
+        result = 1;
+    }
+    else if (version >= OPENSSL_VERSION_1_0_2_RTM)
+    {
+#ifndef SSL_CTRL_SET_ECDH_AUTO
+#define SSL_CTRL_SET_ECDH_AUTO 94
 #endif
+        // Expanded form of SSL_CTX_set_ecdh_auto(ctx, 1)
+        result = SSL_CTX_ctrl(ctx, SSL_CTRL_SET_ECDH_AUTO, 1, NULL);
+    }
+    else
+    {
+        EC_KEY *ecdh = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+
+        if (ecdh != NULL)
+        {
+            result = SSL_CTX_set_tmp_ecdh(ctx, ecdh);
+            EC_KEY_free(ecdh);
+        }
+    }
 
+    // TODO free(version);
```

Comment:
uint32_t-values usually don't need to be freed :smile:

Reply:
Sorry, this was supposed to be `free(version_string)`. See my question on `CryptoNative_SSLEayVersion()`